i6_models/primitives/specaugment.py

```diff
@@ -0,0 +1,123 @@
+import torch
+
+
+def _mask(tensor: torch.Tensor, batch_axis: int, axis: int, pos: torch.Tensor, max_len: int) -> torch.Tensor:
+    """
+    :param tensor: e.g. [B, ..., A, ...] but arbitrary axis order
+    :param batch_axis: index of the batch axis
+    :param axis: which axis A to mask
+    :param pos: at which positions along axis to start the mask (size [B])
+    :param max_len: mask length drawn uniformly from [0, max_len]
+    """
+    batch_dim = tensor.shape[batch_axis]
+    dim = tensor.shape[axis]
+    amount = torch.randint(low=1, high=max_len + 1, size=(batch_dim,), dtype=torch.int32, device=tensor.device)
+    pos2 = torch.min(pos + amount, torch.tensor([dim] * batch_dim, device=tensor.device))
+    idxs = torch.arange(0, dim, device=tensor.device).unsqueeze(0)  # [1,dim]
+    pos_bc = pos.unsqueeze(1)  # [B,1]
+    pos2_bc = pos2.unsqueeze(1)  # [B,1]
+    cond = torch.logical_and(torch.greater_equal(idxs, pos_bc), torch.less(idxs, pos2_bc))  # [B,dim]
+    if batch_axis > axis:
+        cond = cond.transpose(0, 1)  # [dim,B]
+    cond = torch.reshape(
+        cond, shape=[tensor.shape[i] if i in (batch_axis, axis) else 1 for i in range(len(tensor.shape))]
+    )
+    tensor = torch.where(cond, 0.0, tensor)
+    return tensor
+
+
+def _random_mask(tensor: torch.Tensor, batch_axis: int, axis: int, min_num: int, max_num: int, max_len: int):
+    """
+    Mask tensor along axis using N in [min_num, max_num] masks of length [0, max_len]
+
+    :param tensor: e.g. [B, ..., A, ...] but arbitrary axis order
+    :param batch_axis: index of the batch axis
+    :param axis: which axis to mask
+    :param min_num: minimum number of masks
+    :param max_num: maximum number of masks
+    :param max_amount: mask length drawn uniformly from [0, max_amount]
+    """
+
+    batch_dim = tensor.shape[batch_axis]
+    num_masks = torch.randint(min_num, max_num, size=(batch_dim,), device=tensor.device)  # [B]
```

Comment:
I didn't check the old implementation again, I just assumed that this was a replication.

Reply:
I do not understand what you are saying @SimBe195. It is true that the maximum number of masks is determined by the longest sequence in the batch, so a shorter sequence might get more masks if there is another long sequence in the batch. But the average will stay the same, because a mask can also be put into the padding area. This means that in the "expected" case nothing changes.

Reply:
Currently, each sequence in the batch has `num_masks.max()` masks applied to it which is wrong and as it turns out also different from the original tensorflow implementation, so this must be fixed. This is addressed in [this thread](https://github.com/rwth-i6/i6_models/pull/34/files#r1318537237).

Reply:
@SimBe195 So this is resolved?